### ISSUE TITLE
Unsupervised cbow or skipgram of a medium sized corpus slow and memory hungry

### ISSUE THREAD

**Issue by DavidGOrtega** [sszuev/fastText_java#10]
Hi, trying it to generate embeddings of a medium sized corpus like http://mattmahoney.net/dc/enwik8.zip does not train with less 12g of memory, aside of that the speed is 9 words/sec resulting in thousands of hours, while C++ implementation does it in 10min.

Have you ever tried to train such a big corpus?

**Comment by sszuev** [comment]
The provided archive contains data with size ~ 100mb, this is small data.
At the time of developing the process speed and memory consumption were comparable with c++ version.

Please answer: are you using this cloned version or the released one (i.e. v1.0.0) from maven?
In the first case it could be a regression.
Also need to know c++ version.
And environment, including java ver, command line options, etc.
And it would be nice if you compare it with the c++ version that is mentioned in README - this version is a port of the fasttext v0.1.0, not the modern

**Comment by DavidGOrtega** [comment]
@sszuev 

> are you using this cloned version or the released one (i.e. v1.0.0) from maven?

Cloned

> environment

```
MacOS Catalina
MacBook Pro (13-inch, 2018, Four Thunderbolt 3 Ports)
2,3 GHz Intel Core i5
16GB
```

> ncluding java ver

```
java version "1.8.0_202"
Java(TM) SE Runtime Environment (build 1.8.0_202-b08)
Java HotSpot(TM) 64-Bit Server VM (build 25.202-b08, mixed mode)
```
> command line options

```
-Xmx12g
```

 > c++ version

```
master
```
> And it would be nice if you compare it with the c++ version that is mentioned in README

It takes a couple of hours to train the same dataset with the same config than the C++ version, however, with the same configuration, the results are not good, there is something wrong.

> this version is a port of the fasttext v0.1.0, not the modern

You mean training? I have seen that you have updated the original repo to support modern model version including quantization.



I have also a fork of  [fastText_java](https://github.com/ivanhk/fastText_java) its amazing how close we have you and me did the same. Check [my repo](https://github.com/searchbox-ai/fastText_java). I have put also Quantization support and reorganised the code. They look similar!. (WIP)
Man if I had discovered it before I would not have worked in quantization etc... 😬We should work together.

I was working to fix the training code since I had OOM because of the BufferedLineReader when coincidentally I found yours. My fork is slow and memory hungry because of the LineReaders.
Yours is faster but still not comparable with C++

**Comment by sszuev** [comment]
> Cloned

It is quit important if you could confirm that the official released fasttext_java (v1.0.0) has no such dramatic lack of performance. In this case, it can be a regression by someone else, and it can be easily fixed by rollback that commit. Although that means that a bug for which that commit is intended, will be returned back -- this is lesser evil: many people also prefer cloned version, than released.
Also, it can be a bug of the c++ version (0.1.0), which migrated during porting -- sometimes it is difficult to understand what the c++ developers meant due to generic  understanding of OOP and coding style, which are characteristics of c++ developers - that's why need to compare java_v1.0.0 and cpp_v0.1.0.
And, of course, it can be my personal bug, which has been living here starting from the beginning two years ago.
Behaviour of this port and the original c++(v0.1.0) must be comparable, 
and it was, when I worked closely on this repo.  


> -Xmx12g

this is JVM options, it would be nice to have also a program options

> You mean training? I have seen that you have updated the original repo to support modern model version including quantization.

This repo only reflects the c++ version with all its aspects, that is mentioned in README (i.e. v.0.1.0), including training, quantization, etc.
It does not reflect any other changes in the original facebook codebase that were performed after 09/12/2017.

> They look similar!. 

Not surprising - there a lot of ways how to do wrong, but not so much how to do right.

Also please note, there is a network tab with all activities: https://github.com/sszuev/fastText_java/network
@y-yammt has already started to update codebase to modern version, but, it seems, did not complete his changes.


Unfortunately, right now I have no time to work on this repo closely, it requires active coding and deep dive; but now I can only fix obvious things, or accept pull requests.
However, I believe that Java might offer much more advanced facilities to control, develop and speedup things, than C++. 
For example, Java9 compressed `String`s might reduce memory consumption, fork-joint and parallel processing might speedup the calculation (at least for machines with many CPUs). 
Also, there were thoughts about hadoop and map-reduce.

And if you would like to work seriously on this repo, for example investigate and fix this problem, that would be great. 
I'd also like to continue, someday, not right now.
